### PR TITLE
imx6 enet with TCP_WRITE_BUFFERS

### PR DESCRIPTION
**Pull request by masayuki2009**
## Summary

- This PR consists of the following 3 commits
- commit 1: boards: sabre-6quad: the iperf works with CONFIG_NET_TCP_WRITE_BUFFERS
   - This commit makes the iperf work with CONFIG_NET_TCP_WRITE_BUFFERS
   - Since the iperf sends a large packet (e.g. 16KB), CONFIG_IOB_NBUFFERS
      must be increased enough to hold the packet
   - To improve performance CONFIG_IMX_ENET_NTXBUFFERS was changed to
      the default (i.e. 2)
- commit 2: net: tcp: Add DEBUGASSERT() in psock_tcp_send()
  - This commit adds DEBUGASSERT() to check the IOB size
- commit 3: arch: imx6: Add CONFIG_IMX_ENET_NTXBUFFERS check in imx_enet.c
  -  This commit checks CONFIG_IMX_ENET_NTXBUFFERS without
     CONFIG_NET_TCP_WRITE_BUFFERS

## Impact

- sabre-6quad:netnsh

## Testing

- Tested with iperf with QEMU

### CODE REVIEW

**Review by xiaoxiang781216 — APPROVED** [review]

File: net/tcp/tcp_send_buffered.c [hunk 1]
```diff
@@ -1144,6 +1144,11 @@ ssize_t psock_tcp_send(FAR struct socket *psock, FAR const void *buf,
            */
 
           blresult = net_breaklock(&count);
+
+          /* NOTE: At least IOB needs to hold the packet */
+
+          DEBUGASSERT((CONFIG_IOB_NBUFFERS * CONFIG_IOB_BUFSIZE) > len);
```
Comment (yamt) [inline net/tcp/tcp_send_buffered.c:1150]: panicking on user-supplied value (`len`) is not a good idea.
what's wrong with making the below TCP_WBCOPYIN fail?

File: arch/arm/src/imx6/imx_enet.c [hunk 2]
```diff
@@ -102,6 +102,10 @@
 #  error "Need at least one TX buffer"
 #endif
 
+#if !defined (CONFIG_NET_TCP_WRITE_BUFFERS) && (CONFIG_IMX_ENET_NTXBUFFERS != 1)
+#  error "CONFIG_IMX_ENET_NTXBUFFERS must be 1 without TCP_WRITE_BUFFERS"
```
Comment (davids5) [inline arch/arm/src/imx6/imx_enet.c:106]: error "TCP_WRITE_BUFFERS are disabled. CONFIG_IMX_ENET_NTXBUFFERS must be set to 1"

Comment (masayuki2009) [reply arch/arm/src/imx6/imx_enet.c:106]: @davids5 

Thanks! I'll send a new PR later.

File: net/tcp/tcp_send_buffered.c (hunk 1 above)
Comment (masayuki2009) [reply net/tcp/tcp_send_buffered.c:1150]: @yamt 

Actually, TCP_WBCOPYIN() does not finish.
I confirmed it with CONFIG_IOB_DEBUG=y 

```
        Interval Bandwidth 

[ 25.260000] psock_tcp_send: len=16384 
[ 25.260000] iob_copyin_internal: iob=0x1084754c len=16384 offset=0 
[ 25.260000] iob_copyin_internal: iob=0x1084754c avail=0 len=16384 next=0 
[ 25.260000] iob_copyin_internal: iob=0x1084754c Copy 196 bytes new len=196 
[ 25.260000] iob_copyin_internal: iob=0x1084754c added to the chain 
[ 25.260000] iob_copyin_internal: iob=0x10847480 avail=0 len=16188 next=0 
[ 25.260000] iob_copyin_internal: iob=0x10847480 Copy 196 bytes new len=196 
[ 25.260000] iob_copyin_internal: iob=0x10847480 added to the chain 
[ 25.260000] iob_copyin_internal: iob=0x108473b4 avail=0 len=15992 next=0 
[ 25.260000] iob_copyin_internal: iob=0x108473b4 Copy 196 bytes new len=196 
[ 25.260000] iob_copyin_internal: iob=0x108473b4 added to the chain 
[ 25.260000] iob_copyin_internal: iob=0x108472e8 avail=0 len=15796 next=0 
[ 25.260000] iob_copyin_internal: iob=0x108472e8 Copy 196 bytes new len=196 
[ 25.270000] iob_copyin_internal: iob=0x108472e8 added to the chain 
[ 25.270000] iob_copyin_internal: iob=0x1084721c avail=0 len=15600 next=0 
[ 25.270000] iob_copyin_internal: iob=0x1084721c Copy 196 bytes new len=196 
[ 25.270000] iob_copyin_internal: iob=0x1084721c added to the chain 
[ 25.270000] iob_copyin_internal: iob=0x10847150 avail=0 len=15404 next=0 
[ 25.270000] iob_copyin_internal: iob=0x10847150 Copy 196 bytes new len=196 
[ 25.270000] iob_copyin_internal: iob=0x10847150 added to the chain 
[ 25.270000] iob_copyin_internal: iob=0x10847084 avail=0 len=15208 next=0 
[ 25.270000] iob_copyin_internal: iob=0x10847084 Copy 196 bytes new len=196 
[ 25.270000] iob_copyin_internal: iob=0x10847084 added to the chain 
[ 25.270000] iob_copyin_internal: iob=0x10846fb8 avail=0 len=15012 next=0 
[ 25.270000] iob_copyin_internal: iob=0x10846fb8 Copy 196 bytes new len=196 
[ 25.270000] iob_copyin_internal: iob=0x10846fb8 added to the chain 
[ 25.270000] iob_copyin_internal: iob=0x10846eec avail=0 len=14816 next=0 
[ 25.270000] iob_copyin_internal: iob=0x10846eec Copy 196 bytes new len=196 
[ 25.270000] iob_copyin_internal: iob=0x10846eec added to the chain 
[ 25.270000] iob_copyin_internal: iob=0x10846e20 avail=0 len=14620 next=0 
[ 25.270000] iob_copyin_internal: iob=0x10846e20 Copy 196 bytes new len=196 
[ 25.270000] iob_copyin_internal: iob=0x10846e20 added to the chain 
[ 25.270000] iob_copyin_internal: iob=0x10846d54 avail=0 len=14424 next=0 
[ 25.270000] iob_copyin_internal: iob=0x10846d54 Copy 196 bytes new len=196 
[ 25.270000] iob_copyin_internal: iob=0x10846d54 added to the chain 
[ 25.270000] iob_copyin_internal: iob=0x10846c88 avail=0 len=14228 next=0 
[ 25.270000] iob_copyin_internal: iob=0x10846c88 Copy 196 bytes new len=196 
[ 25.270000] iob_copyin_internal: iob=0x10846c88 added to the chain 
[ 25.270000] iob_copyin_internal: iob=0x10846bbc avail=0 len=14032 next=0 
[ 25.270000] iob_copyin_internal: iob=0x10846bbc Copy 196 bytes new len=196 
[ 25.270000] iob_copyin_internal: iob=0x10846bbc added to the chain 
[ 25.270000] iob_copyin_internal: iob=0x10846af0 avail=0 len=13836 next=0 
[ 25.270000] iob_copyin_internal: iob=0x10846af0 Copy 196 bytes new len=196 
[ 25.270000] iob_copyin_internal: iob=0x10846af0 added to the chain 
[ 25.270000] iob_copyin_internal: iob=0x10846a24 avail=0 len=13640 next=0 
[ 25.270000] iob_copyin_internal: iob=0x10846a24 Copy 196 bytes new len=196 
[ 25.270000] iob_copyin_internal: iob=0x10846a24 added to the chain 
[ 25.270000] iob_copyin_internal: iob=0x10846958 avail=0 len=13444 next=0 
[ 25.270000] iob_copyin_internal: iob=0x10846958 Copy 196 bytes new len=196 
[ 25.270000] iob_copyin_internal: iob=0x10846958 added to the chain 
[ 25.270000] iob_copyin_internal: iob=0x1084688c avail=0 len=13248 next=0 
[ 25.280000] iob_copyin_internal: iob=0x1084688c Copy 196 bytes new len=196 
[ 25.280000] iob_copyin_internal: iob=0x1084688c added to the chain 
[ 25.280000] iob_copyin_internal: iob=0x108467c0 avail=0 len=13052 next=0 
[ 25.280000] iob_copyin_internal: iob=0x108467c0 Copy 196 bytes new len=196 
[ 25.280000] iob_copyin_internal: iob=0x108467c0 added to the chain 
[ 25.280000] iob_copyin_internal: iob=0x108466f4 avail=0 len=12856 next=0 
[ 25.280000] iob_copyin_internal: iob=0x108466f4 Copy 196 bytes new len=196 
[ 25.280000] iob_copyin_internal: iob=0x108466f4 added to the chain 
[ 25.280000] iob_copyin_internal: iob=0x10846628 avail=0 len=12660 next=0 
[ 25.280000] iob_copyin_internal: iob=0x10846628 Copy 196 bytes new len=196 
[ 25.280000] iob_copyin_internal: iob=0x10846628 added to the chain 
[ 25.280000] iob_copyin_internal: iob=0x1084655c avail=0 len=12464 next=0 
[ 25.280000] iob_copyin_internal: iob=0x1084655c Copy 196 bytes new len=196 
[ 25.280000] iob_copyin_internal: iob=0x1084655c added to the chain 
[ 25.280000] iob_copyin_internal: iob=0x10846490 avail=0 len=12268 next=0 
[ 25.280000] iob_copyin_internal: iob=0x10846490 Copy 196 bytes new len=196 
[ 25.280000] iob_copyin_internal: iob=0x10846490 added to the chain 
[ 25.280000] iob_copyin_internal: iob=0x108463c4 avail=0 len=12072 next=0 
[ 25.280000] iob_copyin_internal: iob=0x108463c4 Copy 196 bytes new len=196 
[ 25.280000] iob_copyin_internal: iob=0x108463c4 added to the chain 
[ 25.280000] iob_copyin_internal: iob=0x108462f8 avail=0 len=11876 next=0 
[ 25.280000] iob_copyin_internal: iob=0x108462f8 Copy 196 bytes new len=196 
[ 25.280000] iob_copyin_internal: iob=0x108462f8 added to the chain 
[ 25.280000] iob_copyin_internal: iob=0x1084622c avail=0 len=11680 next=0 
[ 25.280000] iob_copyin_internal: iob=0x1084622c Copy 196 bytes new len=196 
[ 25.280000] iob_copyin_internal: iob=0x1084622c added to the chain 
[ 25.280000] iob_copyin_internal: iob=0x10846160 avail=0 len=11484 next=0 
[ 25.280000] iob_copyin_internal: iob=0x10846160 Copy 196 bytes new len=196 
[ 25.280000] iob_copyin_internal: iob=0x10846160 added to the chain 
[ 25.280000] iob_copyin_internal: iob=0x10846094 avail=0 len=11288 next=0 
[ 25.280000] iob_copyin_internal: iob=0x10846094 Copy 196 bytes new len=196 
[ 25.280000] iob_copyin_internal: iob=0x10846094 added to the chain 
[ 25.280000] iob_copyin_internal: iob=0x10845fc8 avail=0 len=11092 next=0 
[ 25.280000] iob_copyin_internal: iob=0x10845fc8 Copy 196 bytes new len=196 
[ 25.280000] iob_copyin_internal: iob=0x10845fc8 added to the chain 
[ 25.280000] iob_copyin_internal: iob=0x10845efc avail=0 len=10896 next=0 
[ 25.280000] iob_copyin_internal: iob=0x10845efc Copy 196 bytes new len=196 
[ 25.280000] iob_copyin_internal: iob=0x10845efc added to the chain 
[ 25.280000] iob_copyin_internal: iob=0x10845e30 avail=0 len=10700 next=0 
[ 25.290000] iob_copyin_internal: iob=0x10845e30 Copy 196 bytes new len=196 
[ 25.290000] iob_copyin_internal: iob=0x10845e30 added to the chain 
[ 25.290000] iob_copyin_internal: iob=0x10845d64 avail=0 len=10504 next=0 
[ 25.290000] iob_copyin_internal: iob=0x10845d64 Copy 196 bytes new len=196 
[ 25.290000] iob_copyin_internal: iob=0x10845d64 added to the chain 
[ 25.290000] iob_copyin_internal: iob=0x10845c98 avail=0 len=10308 next=0 
[ 25.290000] iob_copyin_internal: iob=0x10845c98 Copy 196 bytes new len=196 
[ 25.290000] iob_copyin_internal: iob=0x10845c98 added to the chain 
[ 25.290000] iob_copyin_internal: iob=0x10845bcc avail=0 len=10112 next=0 
[ 25.290000] iob_copyin_internal: iob=0x10845bcc Copy 196 bytes new len=196 
[ 25.290000] iob_copyin_internal: iob=0x10845bcc added to the chain 
[ 25.290000] iob_copyin_internal: iob=0x10845b00 avail=0 len=9916 next=0 
[ 25.290000] iob_copyin_internal: iob=0x10845b00 Copy 196 bytes new len=196 
[ 25.290000] iob_copyin_internal: iob=0x10845b00 added to the chain 
[ 25.290000] iob_copyin_internal: iob=0x10845a34 avail=0 len=9720 next=0 
[ 25.290000] iob_copyin_internal: iob=0x10845a34 Copy 196 bytes new len=196 
[ 25.290000] iob_copyin_internal: iob=0x10845a34 added to the chain 
[ 25.290000] iob_copyin_internal: iob=0x10845968 avail=0 len=9524 next=0 
[ 25.290000] iob_copyin_internal: iob=0x10845968 Copy 196 bytes new len=196 
[ 26.260000] tcp_callback: flags: 0010 
[ 26.260000] psock_send_eventhandler: flags: 0010 
[ 26.260000] tcp_appsend: result: 0010 d_sndlen: 0 conn->tx_unacked: 0 
[ 26.260000] tcp_rexmit: result: 0010 d_sndlen: 0 conn->tx_unacked: 0 
```

Comment (masayuki2009) [reply net/tcp/tcp_send_buffered.c:1150]: @yamt 

Here is the call stack.

```
(gdb) where                                                                                                                                                                                                                                          
#0  arm_saveusercontext () at armv7-a/arm_saveusercontext.S:129                                                                                                                                                                                      
#1  0x108080cc in up_block_task (tcb=tcb@entry=0x108545f0, task_state=task_state@entry=TSTATE_WAIT_SEM) at armv7-a/arm_blocktask.c:147                                                                                                               
#2  0x108019d8 in nxsem_wait (sem=sem@entry=0x1084b760 <g_iob_sem>) at semaphore/sem_wait.c:153                                                                                                                                                      
#3  0x10801a00 in nxsem_wait_uninterruptible (sem=sem@entry=0x1084b760 <g_iob_sem>) at semaphore/sem_wait.c:222                                                                                                                                      
#4  0x10824524 in iob_allocwait (consumerid=IOBUSER_NET_TCP_WRITEBUFFER, throttled=0 '\000') at iob/iob_alloc.c:133                                                                                                                                  
#5  iob_alloc (throttled=<optimized out>, consumerid=<optimized out>) at iob/iob_alloc.c:204                                                                                                                                                         
#6  0x10824730 in iob_copyin_internal (iob=0x10844b18 <g_iob_pool>, src=0x10852170 "", len=9328, offset=<optimized out>, offset@entry=0, throttled=throttled@entry=0 '\000', can_block=can_block@entry=1 '\001', consumerid=IOBUSER_NET_TCP_WRITEBUF\
FER, consumerid@entry=40) at iob/iob_copyin.c:180                                                                                                                                                                                                    
#7  0x108247a8 in iob_copyin_internal (consumerid=40, can_block=1 '\001', throttled=0 '\000', offset=offset@entry=0, len=len@entry=16384, src=src@entry=0x108505e0 "", iob=<optimized out>) at iob/iob_copyin.c:72                                   
#8  iob_copyin (iob=<optimized out>, src=src@entry=0x108505e0 "", len=len@entry=16384, offset=offset@entry=0, throttled=throttled@entry=0 '\000', consumerid=consumerid@entry=IOBUSER_NET_TCP_WRITEBUFFER) at iob/iob_copyin.c:223                   
#9  0x10830228 in psock_tcp_send (psock=<optimized out>, buf=0x108505e0, len=<optimized out>, flags=<optimized out>) at tcp/tcp_send_buffered.c:1151                                                                                                 
#10 0x10812954 in psock_send (psock=<optimized out>, buf=buf@entry=0x108505e0, len=len@entry=16384, flags=<optimized out>) at socket/send.c:90                                                                                                       
#11 0x10812984 in nx_send (sockfd=sockfd@entry=3, buf=buf@entry=0x108505e0, len=len@entry=16384, flags=<optimized out>) at socket/send.c:131                                                                                                         
#12 0x10812990 in send (sockfd=sockfd@entry=3, buf=buf@entry=0x108505e0, len=len@entry=16384, flags=<optimized out>) at socket/send.c:208                                                                                                            
#13 0x10828434 in iperf_run_tcp_client () at iperf.c:562                                                                                                                                                                                             
#14 iperf_task_traffic (arg=<optimized out>) at iperf.c:600                                                                                                                                                                                          
#15 0x1081ca80 in pthread_start () at pthread/pthread_create.c:183      
```

Comment (yamt) [reply net/tcp/tcp_send_buffered.c:1150]: @masayuki2009
we can make it bail out earlier, or improve TCP_WBCOPYIN.
or maybe we can make it send the data chunk-by-chunk.
my point is that DEBUGASSERT is not a right think to do here.

Comment (masayuki2009) [reply net/tcp/tcp_send_buffered.c:1150]: @yamt 

OK, I will send a new PR to remove the DEBUGASSERT() for now.

**Pull request merged**